### DISCUSSION TITLE
Support for Switch Expressions on Maybe that uses Property Pattern matching

### DISCUSSION THREAD

**Discussion by julienasp** [vkhorikov/CSharpFunctionalExtensions#559] [Ideas]
@vkhorikov  today at work, one of my colleague used a switch expression on a maybe. 

and he used it via the Property Pattern matching and not the positional pattern + deconstruct.

it looked something like this:
![image](https://github.com/user-attachments/assets/24a05d67-d731-44df-b675-bdea32cb6316)

and because of the current implementation of the Value property is was breaking as expected
![image](https://github.com/user-attachments/assets/07a473f8-4810-47e9-885b-78ae3de58ecb)

so every time you wish you use switch expression with property matching it would fail if you try to switch on the Value property value.

so the only way to use a switch expression with the Maybe type is via the positional pattern matching that uses the deconstruct like this
![image](https://github.com/user-attachments/assets/012cb9c8-4533-489c-a782-2be9c353e8fb)
or
with the Property Matching with some ordering like this, to validate the HasValue first
![image](https://github.com/user-attachments/assets/2b0a8dd2-dc97-4721-91fd-21e5810e2db2)

or be robust upfront with a If, or use some safety execution extension like the Maybe.Execute

so i was thinking, okay i'll push some changes, to simply return the _value instead of GetValueOrThrow but that's a massive breaking change.

It would be nice to have support that feature out of the box, but it implies a huge behaviour change.

**Comment by vkhorikov** [comment]
Yeah, it would be nice, but only as long as the behavior of `Value` being unavailable stays in place. Failing fast is a big reason why `Maybe` was developed in the first place. Not sure if this is going to be possible.

This solution looks good, though?

```cs
return switch
{
  { HasNoValue: true } => "a",
  { Value: "X" } => "b"
}
```

This is the same as:

```cs
if (maybe.HasNoValue)
  return "a";

if (maybe.Value == "X")
  return "b";
```
